openflow13/nxext_test.go

```diff
@@ -0,0 +1,215 @@
+		t.Errorf("Failed to decode ofs, expect: %d, actual: %d", ofs, decodeOfs)
+	}
+	if nBits != deCodeNbits {
+		t.Errorf("Failed to decode nBits, expect: %d, actual: %d", nBits, deCodeNbits)
+	}
+}
+
+func TestUint32Message(t *testing.T) {
+	tgtData := uint32(0xff00ff00)
+	msg := newUint32Message(tgtData)
+	if msg.Len() != 4 {
+		t.Errorf("Failed to generate Uint32Message")
+	}
+	testData, err := msg.MarshalBinary()
+	if err != nil {
+		t.Errorf("Failed to invoke Uint32Message.MarshalBinary, %v", err)
+	}
+	var testMsg = new(Uint32Message)
+	err = testMsg.UnmarshalBinary(testData)
+	if err != nil {
+		t.Errorf("Failed to invoke Uint32Message.UnmarshalBinary, %v", err)
+	}
+	if testMsg.data != tgtData {
+		t.Errorf("Failed to recover uint32 from Uint32Message, tgt: %d, actual: %d", tgtData, testMsg.data)
+	}
+}
+
+func TestUint16Message(t *testing.T) {
+	tgtData := uint16(0xfe10)
+	msg := newUint16Message(tgtData)
+	if msg.Len() != 2 {
+		t.Errorf("Failed to generate Uint32Message")
+	}
+	testData, err := msg.MarshalBinary()
+	if err != nil {
+		t.Errorf("Failed to invoke Uint32Message.MarshalBinary, %v", err)
+	}
+	var testMsg = new(Uint16Message)
+	err = testMsg.UnmarshalBinary(testData)
+	if err != nil {
+		t.Errorf("Failed to invoke Uint32Message.UnmarshalBinary, %v", err)
+	}
+	if testMsg.data != tgtData {
+		t.Errorf("Failed to recover uint32 from Uint32Message, tgt: %d, actual: %d", tgtData, testMsg.data)
+	}
+}
+
+func TestNewUintXMask(t *testing.T) {
+	tgtData32 := uint32(0xff)
+	rng1 := &NXRange{start: 0, end: 7}
+	testMsg := &Uint32Message{data: rng1.ToUint32Mask()}
+	if testMsg.data != tgtData32 {
+		t.Errorf("Failed to invoke newUint32Mask, expected: %02x, actual: %02x", tgtData32, testMsg.data)
+	}
+}
+
+func TestNXMFieldHeader(t *testing.T) {
+	for k := range oxxFieldHeaderMap {
+		field1, err := FindFieldHeaderByName(k, false)
+		if err != nil {
+			t.Errorf("Test failed: %v", err)
+		}
+		testMatchFieldHeaderMarshalUnMarshal(field1, t)
+		field2, _ := FindFieldHeaderByName(k, true)
+		testMatchFieldHeaderMarshalUnMarshal(field2, t)
+	}
+}
+
+func TestCTLabel(t *testing.T) {
+	var label = [16]byte{}
+	testData, err := hex.DecodeString(fmt.Sprintf("%d", 0x12345678))
+	copy(label[:], testData)
+	field := newCTLabel(label)
+	data, err := field.MarshalBinary()
+	if err != nil {
+		t.Errorf("Failed to MarshalBinary CTLabel: %v", err)
+	}
+	field2 := new(CTLabel)
+	field2.UnmarshalBinary(data)
+	if field2.data != label {
+		t.Errorf("Failed to UnmarshalBinary CTLabel: %v", err)
+	}
+
+	var mask = [16]byte{}
+	binary.BigEndian.PutUint32(mask[:], 0xffffffff)
+}
+
+func testMatchFieldHeaderMarshalUnMarshal(tgtField *MatchField, t *testing.T) {
+	headerInt := tgtField.MarshalHeader()
+	testMFHeader := new(MatchField)
+	var data []byte = make([]byte, 4)
+	binary.BigEndian.PutUint32(data, headerInt)
+	_ = testMFHeader.UnmarshalHeader(data)
+	if testMFHeader.Class != tgtField.Class {
+		t.Errorf("Failed to UnmarshalHeader: class")
```

Comment:
that's a weird error message: you didn't fail to unmarshal, the result is just not correct

 Did you mean something like "Unmarshalled header has incorrect 'Class' field"?

Reply:
Yes, you are correct. I would change as you suggested.